src/app/functions/server/getRelevantGuidedSteps.ts

```diff
@@ -0,0 +1,260 @@
+    id: "StartScan",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/data-broker-profiles/view-data-brokers",
+    id: "ScanResult",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/social-security-number",
+    id: "HighRiskSsn",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/credit-card-number",
+    id: "HighRiskCreditCard",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/bank-account",
+    id: "HighRiskBankAccount",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/high-risk-data-breaches/pin-number",
+    id: "HighRiskPin",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/leaked-passwords/password",
+    id: "LeakedPasswordsPassword",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/leaked-passwords/security-question",
+    id: "LeakedPasswordsSecurityQuestion",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/phone",
+    id: "SecurityTipsPhone",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/email",
+    id: "SecurityTipsEmail",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/ip",
+    id: "SecurityTipsIp",
+  },
+  {
+    href: "/redesign/user/dashboard",
+    id: "Done",
+  },
+] as const satisfies ReadonlyArray<{ href: string; id: string }>;
+
+export type StepLink = (typeof stepLinks)[number];
+export type StepLinkWithStatus = (typeof stepLinks)[number] & {
+  eligible: boolean;
+  completed: boolean;
+};
+
+export type OutputData = {
+  current: StepLink | null;
+  skipTarget: StepLink | null;
+};
+
+export function getRelevantGuidedSteps(
+  data: InputData,
+  fromStep?: StepLink["id"]
+): OutputData {
+  const current = getNextGuidedStep(data, fromStep);
+
+  // There should be a relevant next step for every user (even if it's just
+  // going back to the dashbord), so we can't hit this line in tests (and
+  // shouldn't be able to in production either):
+  /* c8 ignore next 6 */
+  if (current === null) {
+    return {
+      current: null,
+      skipTarget: null,
+    };
+  }
+
+  return {
+    current: current,
+    skipTarget: getNextGuidedStep(data, current.id),
+  };
+}
+
+function getNextGuidedStep(
+  data: InputData,
+  fromStep?: StepLink["id"]
+): StepLink | null {
+  // Resisting the urge to add a state machine... ^.^
+  const stepLinkStatuses = getGuidedStepStatuses(data);
+  const fromIndex = stepLinkStatuses.findIndex((step) => step.id === fromStep);
+  const stepLinksToCheck =
+    fromIndex !== -1 ? stepLinkStatuses.slice(fromIndex) : stepLinkStatuses;
+  const nextStep = stepLinksToCheck.find((stepLink) => {
+    return stepLink.eligible && !stepLink.completed;
+  });
+
+  // There should be a relevant next step for every user (even if it's just
+  // going back to the dashbord), so we can't hit this line in tests (and
+  // shouldn't be able to in production either):
```

Comment:
Is it worth it to add Glean pings or server logs in the future to make sure dead code paths are really not being hit in production? This could help us detect if there is a significant number of people who unintentionally speed-run the breach resolution.

Reply:
Yeah good idea, added in 804575af0.

Also made me realise that this specific line should actually be coverable in tests, since it's also called to determine the _next_ step - and when the user is done, there might not be one.

Which, in turn, made me realise that there was actually a bug: when calculating the next relevant step, instead of evaluating all steps after the current step, I was evaluating all steps _starting with_ the current step - which would, of course, always match the current step. So that is now fixed too: 83669d8e4